test/e2e/storage/sanity.go

```diff
@@ -996,3 +996,36 @@ func WaitForPodsWithLabelRunningReady(c clientset.Interface, ns string, label la
 		})
 	return pods, err
 }
+
+// We register list of volumes before test, using out-of-band commands (i.e. not CSI API but direct host commands).
+// checkForLeftovers lists volumes again after test, and comparing shows possible leftovers.
+// We avoid complexity of tracking nodes separately, by combining all outputs into same string.
+// TODO: remove debug prints before merging
+func getVolumesList(numWorkers int) string {
+	var cmd string
+	var buf string
+	switch os.Getenv("TEST_DEVICEMODE") {
+	case "direct":
+		cmd = "sudo ndctl list"
+	case "lvm":
+		cmd = "sudo lvs --foreign --noheadings"
+	}
+	framework.Logf("getVolumesList-DEBUG: cmd:[%s]", cmd)
+	for worker := 1; worker <= numWorkers; worker++ {
+		sshcmd := fmt.Sprintf("%s/_work/%s/ssh.%d", os.Getenv("REPO_ROOT"), os.Getenv("CLUSTER"), worker)
+		ssh := exec.Command(sshcmd, cmd)
+		out, _ := ssh.CombinedOutput()
+		//framework.Logf("get VolumeList-DEBUG(%d): cmd:[%s] out:[%s]", worker, cmd, out)
+		nodeheader := fmt.Sprintf("=== Volumes on Node %d:\n", worker)
+		buf += nodeheader
+		buf += string(out)
+	}
+	framework.Logf("getVolumeList-DEBUG: combined:[%s]", buf)
+	return buf
+}
+
+func checkForLeftoverVolumes(numWorkers int, volBefore string) {
+	volNow := getVolumesList(numWorkers)
+	framework.Logf("checkForLeftoverVolumes-DEBUG: now:[%s] before:[%s]", volNow, volBefore)
+	Expect(volNow).To(Equal(volBefore), "same volumes before and after the test")
```

Comment:
The log output does not help when the comparison fails because it is printed separately from the test failure:
```
[2019-11-04T13:24:50.878Z]   /go/src/github.com/intel/pmem-csi/test/e2e/storage/sanity.go:168
[2019-11-04T13:24:50.878Z] Nov  4 13:24:50.545: INFO: getVolumesList-DEBUG: cmd:[sudo lvs --foreign --noheadings]
[2019-11-04T13:24:51.641Z] Nov  4 13:24:51.500: INFO: getVolumeList-DEBUG: combined:[=== Volumes on Node 1:
[2019-11-04T13:24:51.641Z] === Volumes on Node 2:
[2019-11-04T13:24:51.641Z]   6a2130accf186ac16f0d4f336bd488c9a512513f ndbus0region0fsdax -wi-a----- 4.00m                                                    
[2019-11-04T13:24:51.641Z] === Volumes on Node 3:
[2019-11-04T13:24:51.641Z] ]
[2019-11-04T13:24:51.641Z] Nov  4 13:24:51.500: INFO: checkForLeftoverVolumes-DEBUG: now:[=== Volumes on Node 1:
[2019-11-04T13:24:51.641Z] === Volumes on Node 2:
[2019-11-04T13:24:51.641Z]   6a2130accf186ac16f0d4f336bd488c9a512513f ndbus0region0fsdax -wi-a----- 4.00m                                                    
[2019-11-04T13:24:51.641Z] === Volumes on Node 3:
[2019-11-04T13:24:51.641Z] ] before:[=== Volumes on Node 1:
[2019-11-04T13:24:51.641Z] === Volumes on Node 2:
[2019-11-04T13:24:51.641Z] === Volumes on Node 3:
[2019-11-04T13:24:51.641Z] ]
[2019-11-04T13:24:51.641Z] 
[2019-11-04T13:24:51.641Z] [91m[1m• Failure in Spec Teardown (AfterEach) [3.416 seconds][0m
[2019-11-04T13:24:51.641Z] sanity
[2019-11-04T13:24:51.641Z] [90m/go/src/github.com/intel/pmem-csi/test/e2e/storage/sanity.go:55[0m
[2019-11-04T13:24:51.641Z]   [91m[1mController Service [Controller Server] [AfterEach][0m
[2019-11-04T13:24:51.641Z]   [90m/go/src/github.com/intel/pmem-csi/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/tests.go:44[0m
[2019-11-04T13:24:51.641Z]     ListVolumes
[2019-11-04T13:24:51.641Z]     [90m/go/src/github.com/intel/pmem-csi/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:166[0m
[2019-11-04T13:24:51.641Z]       pagination should detect volumes added between pages and accept tokens when the last volume from a page is deleted
[2019-11-04T13:24:51.641Z]       [90m/go/src/github.com/intel/pmem-csi/vendor/github.com/kubernetes-csi/csi-test/pkg/sanity/controller.go:262[0m
[2019-11-04T13:24:51.641Z] 
[2019-11-04T13:24:51.641Z]       [91msame volumes before and after the test
[2019-11-04T13:24:51.641Z]       Expected
[2019-11-04T13:24:51.641Z]           <string>: "...e 2:
[2019-11-04T13:24:51.641Z]             6a21..."
[2019-11-04T13:24:51.641Z]       to equal               |
[2019-11-04T13:24:51.641Z]           <string>: "...e 2:
[2019-11-04T13:24:51.641Z]           === Vo..."[0m
[2019-11-04T13:24:51.641Z] 
[2019-11-04T13:24:51.641Z]       /go/src/github.com/intel/pmem-csi/test/e2e/storage/sanity.go:1030
```

The strings also got truncated.

Please try comparing string lists and put the node number into the `To` description.